### ISSUE TITLE
How to deal with a minion “already deleted from tracker”?

### ISSUE THREAD

**Issue by wsw70** [saltstack/salt#45728]
### Description of Issue/Question
_Initially asked on [SO](https://stackoverflow.com/questions/48457530/how-to-deal-with-a-minion-already-deleted-from-tracker)_

When running 

    salt '*' test.ping -s --out json

I get the following output (reduced to just a few typical cases, the ellipsis covers the same cases)

    Minion sent222 did not respond. No job will be sent.
    (...)
    minion sent005 was already deleted from tracker, probably a duplicate key
    (...)
    {
    (...)
        "sent005": true,
    (...)
    }

The case of `sent222` is clear: it is not available.

What should I do with `sent002`? It does answer, but also warns about being deleted from the tracker and having a duplicate key. **What does this mean?**

I noticed this by chance because the regular `salt '*' test.ping` simply states that `sent005` is `True`, without any further comments on its state.

### Versions Report
```
root@sentmaster ~# salt --versions-report
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.1
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-3-amd64
         system: Linux
        version: debian 9.1
```

**Comment by gtmanfred** [comment]
Looks like this was added here.

```
commit a8ff42593170a951e5ea35a23e7eecb1e88ecee2
Author: Paul Miller <jettero@corky.vhb.voltar.org>
Date:   Wed Mar 9 16:27:52 2016 -0500

    reliably detect duplicate keys during batch

    Sometimes some of the admins rsync copy /etc/salt configurations to new hosts.
    I can detect these duplicates with this edit. The usual methods don't work
    since the key is the same as the other hosts with that id.

diff --git a/salt/cli/batch.py b/salt/cli/batch.py
index 83e44b5370..a114d1c109 100644
--- a/salt/cli/batch.py
+++ b/salt/cli/batch.py
@@ -180,11 +180,17 @@ class Batch(object):
                             continue
                         if self.opts.get('raw'):
                             parts.update({part['id']: part})
-                            minion_tracker[queue]['minions'].remove(part['id'])
+                            if part['id'] in minion_tracker[queue]['minions']:
+                                minion_tracker[queue]['minions'].remove(part['id'])
+                            else:
+                                print_cli('minion {0} was already deleted from tracker, probably a duplicate key'.format(part['id']))
                         else:
                             parts.update(part)
                             for id in part.keys():
-                                minion_tracker[queue]['minions'].remove(id)
+                                if id in minion_tracker[queue]['minions']:
+                                    minion_tracker[queue]['minions'].remove(id)
+                                else:
+                                    print_cli('minion {0} was already deleted from tracker, probably a duplicate key'.format(id))
                 except StopIteration:
                     # if a iterator is done:
                     # - set it to inactive
```

https://github.com/saltstack/salt/pull/31777

Are you rsyncing minion keys between masters?

Thanks,
Daniel

**Comment by wsw70** [comment]
@gtmanfred I have only one master, but who listens on two different ports. It may be that a minion has contacted either of these ports. Following on your comment, I ran

```
root@sentmaster ~# salt '*' test.ping | grep sent | sort
sent002:
sent002:
(...)
```

I do not know salt enough to understand this result. This is ONE minion, who (maybe) connected to either of the ports at some time and is currently up (there were two `True` for each of the two `sent002` minions in the output - again, this is one machine).

**Comment by gtmanfred** [comment]
What happens when you check `ps aux | grep salt-minion` on the sent002 minion?

**Comment by wsw70** [comment]
@gtmanfred : `ps aux | grep salt-minion` shows only one `salt-minion` process

**Comment by wsw70** [comment]
@gtmanfred I restarted them and they do connect to several minions at the same time.

I also have a  case where there are three masters (in the same configuration as above), both the minion and the master was restarted a few times. Several connection still:

```
root@sent003:~# netstat -nn | grep 4505
tcp        0      0 10.0.2.3:33498          10.0.2.1:4505           ESTABLISHED
tcp        0      0 10.130.37.22:38622      10.81.102.133:4505      ESTABLISHED
tcp        0      0 10.130.37.22:45942      192.168.172.98:4505     ESTABLISHED
```

(`10.0.2.3` is a VPN connection, `10.81.102.133 `is a master which holds both the endpoint on its `eth0` and its `tun0` (`10.0.2.3` connects there) and `192.168.172.98` is a second, independent machine)

**Comment by wsw70** [comment]
@gtmanfred I tried to start the minion manually (`salt-minion -l debug`) and I see that it connects to all three masters, in a random way (that is - not according to the list in the config file):

```
[DEBUG   ] Attempting to authenticate with the Salt Master at 10.81.102.133
[INFO    ] Authentication with master at 10.81.102.133 successful!
(this repeated for all three masters)
```

**Comment by gtmanfred** [comment]
Ok, I am unfortunate unable to replicate this.

```
[root@salt ~]# ss -nn | grep 4505
tcp    ESTAB      0      0      172.17.0.2:59216              172.17.0.2:4505
tcp    ESTAB      0      0      172.17.0.2:4505               172.17.0.2:59216
[root@salt ~]# cat /etc/salt/minion
master:
  - 172.17.0.2
  - 127.0.0.1
master_type: failover
master_alive_interval: 30
retry_dns: 0
```

I am seeing the minion only connect once.

```
[root@salt ~]# ss -nn | grep 4505
tcp    ESTAB      0      0      172.17.0.2:59216              172.17.0.2:4505
tcp    ESTAB      0      0      172.17.0.2:4505               172.17.0.2:59216
[root@salt ~]# cat /etc/salt/minion
master:
  - 172.17.0.2
  - 127.0.0.1
master_type: failover
master_alive_interval: 30
retry_dns: 0
```

If i comment out `master_type: failover` then I am seeing this happen, but i am still unable to replicate the initial issue.

Do you have any other minion config files in /etc/salt/minion.d/*.conf on these minions?

Also just to follow up on something that you said earlier

> Ideally followed by "... and the once the first server is back connect to it again"

I do not believe that this will happen, salt will only check that ip again if the one it is connected to fails.

Thanks,
Daniel

**Comment by wsw70** [comment]
I think I found the issue. Out of despair I decided to upgrade the linions - and I discovered that a few of them were at version 2014! I believe that this is the reason they were connecting a few times to the master.

I used the boostrap script to update them (which had all kind of issues as well, specific to raspberry) and now they report only once (checked with `/usr/bin/salt '*' cmd.run 'netstat -an | grep 4505'`).

Thank you for the help and sorry for the trouble of this bug which was not one.
